File: examples/scenarios/checkout_card.py
#!/usr/bin/env python3
"""
Card Payment (Authorize + Capture) - Universal Example

Usage:
    python checkout_card.py --connector=stripe
    python checkout_card.py --connector=adyen
    python checkout_card.py --connector=checkout
"""

import argparse
import asyncio
import json
from pathlib import Path
from typing import Optional

# [START imports]
from google.protobuf.json_format import ParseDict
from payments.generated import sdk_config_pb2, payment_pb2
from payments import PaymentClient
# [END imports]


# [START load_probe_data]
def load_probe_data(connector_name: str) -> dict:
    """Load connector probe data from field_probe."""
    probe_file = Path(__file__).parent.parent.parent / "data" / "field_probe" / f"{connector_name}.json"
    with open(probe_file) as f:
        return json.load(f)
# [END load_probe_data]


# [START stripe_config]
def get_stripe_config(api_key: str) -> sdk_config_pb2.ConnectorConfig:
    """Configuration for Stripe."""
    config = sdk_config_pb2.ConnectorConfig(
        options=sdk_config_pb2.SdkOptions(environment=sdk_config_pb2.Environment.SANDBOX),
    )
    config.connector_config.CopyFrom(payment_pb2.ConnectorSpecificConfig(
        stripe=payment_pb2.StripeConfig(api_key=api_key),
    ))
    return config
# [END stripe_config]


# [START adyen_config]
def get_adyen_config(api_key: str, merchant_account: str) -> sdk_config_pb2.ConnectorConfig:
    """Configuration for Adyen."""
    config = sdk_config_pb2.ConnectorConfig(
        options=sdk_config_pb2.SdkOptions(environment=sdk_config_pb2.Environment.SANDBOX),
    )
    config.connector_config.CopyFrom(payment_pb2.ConnectorSpecificConfig(
        adyen=payment_pb2.AdyenConfig(
            api_key=api_key,
            merchant_account=merchant_account,
        ),
    ))
    return config
# [END adyen_config]


# [START checkout_config]
def get_checkout_config(api_key: str) -> sdk_config_pb2.ConnectorConfig:
    """Configuration for Checkout.com."""
    config = sdk_config_pb2.ConnectorConfig(
        options=sdk_config_pb2.SdkOptions(environment=sdk_config_pb2.Environment.SANDBOX),
    )
    config.connector_config.CopyFrom(payment_pb2.ConnectorSpecificConfig(
        checkout=payment_pb2.CheckoutConfig(api_key=api_key),
    ))
    return config
# [END checkout_config]


# [START build_authorize_request]
def build_authorize_request(probe_data: dict, capture_method: str = "MANUAL") -> dict:
    """Build authorize request from probe data."""
    flows = probe_data.get("flows", {})
    authorize_flows = flows.get("authorize", {})
    
    # Find Card payment method or first supported
    card_data = None
    for pm_key, pm_data in authorize_flows.items():
        if pm_data.get("status") == "supported":
            if pm_key == "Card":
                card_data = pm_data
                break
            elif card_data is None:
                card_data = pm_data
    
    if not card_data:
        raise ValueError("No supported payment method found for authorize flow")
    
    proto_request = dict(card_data.get("proto_request", {}))
    proto_request["capture_method"] = capture_method
    
    return proto_request
# [END build_authorize_request]


# [START build_capture_request]
def build_capture_request(
    connector_transaction_id: str,
    amount: dict,
    merchant_capture_id: str = "capture_001"
) -> dict:
    """Build capture request."""
    return {
        "merchant_capture_id": merchant_capture_id,
        "connector_transaction_id": connector_transaction_id,
        "amount_to_capture": amount,
    }
# [END build_capture_request]


# [START process_checkout_card]
async def process_checkout_card(
    connector_name: str,
    credentials: dict
) -> str:
    """
    Process a card payment with authorize + capture flow.
    
    Args:
        connector_name: Name of the connector (stripe, adyen, etc.)
        credentials: Connector-specific credentials
    
    Returns:
        "success" | "pending" | "failed"
    """
    print(f"\n{'='*60}")
    print(f"Processing Card Payment via {connector_name}")
    print(f"{'='*60}")
    
    # Load probe data for this connector
    probe_data = load_probe_data(connector_name)
    
    # Get configuration
    config = get_connector_config(connector_name, credentials)
    client = PaymentClient(config)
    
    # Build authorize request
    auth_request_dict = build_authorize_request(probe_data, capture_method="MANUAL")
    auth_request = ParseDict(auth_request_dict, payment_pb2.PaymentServiceAuthorizeRequest())
    
    print(f"\nRequest: {json.dumps(auth_request_dict, indent=2)}")
    
    # Step 1: Authorize
    print("\n[1/2] Authorizing...")
    auth_response = await client.authorize(auth_request)
    
    print(f"Status: {auth_response.status}")
    print(f"Connector Transaction ID: {auth_response.connector_transaction_id}")
    
    # Handle status
    if auth_response.status == "FAILED":
        print(f"❌ Payment declined: {auth_response.error_message}")
        return "failed"
    
    if auth_response.status == "PENDING":
        print("⏳ Payment pending - awaiting async confirmation")
        print("   (In production: wait for webhook, then poll get())")
        return "pending"
    
    if auth_response.status != "AUTHORIZED":
        print(f"⚠️  Unexpected status: {auth_response.status}")
        return "error"
    
    print("✅ Funds reserved successfully")
    
    # Step 2: Capture
    print("\n[2/2] Capturing...")
    capture_request_dict = build_capture_request(
        connector_transaction_id=auth_response.connector_transaction_id,
        amount=auth_request_dict["amount"],
    )
    capture_request = ParseDict(capture_request_dict, payment_pb2.PaymentServiceCaptureRequest())
    
    capture_response = await client.capture(capture_request)
    print(f"Capture Status: {capture_response.status}")
    
    if capture_response.status in ["CAPTURED", "AUTHORIZED"]:
        print("✅ Payment captured successfully")
        return "success"
    else:
        print(f"⚠️  Capture returned: {capture_response.status}")
        return "partial"
# [END process_checkout_card]


# [START get_connector_config]
def get_connector_config(connector_name: str, credentials: dict) -> sdk_config_pb2.ConnectorConfig:
    """Get configuration for any supported connector."""
    config_functions = {
        "stripe": lambda: get_stripe_config(credentials.get("api_key")),
        "adyen": lambda: get_adyen_config(
            credentials.get("api_key"),
            credentials.get("merchant_account")
        ),
        "checkout": lambda: get_checkout_config(credentials.get("api_key")),
    }
    
    if connector_name not in config_functions:
        raise ValueError(f"Unknown connector: {connector_name}")
    
    return config_functions[connector_name]()
# [END get_connector_config]


# [START main]
def main():
    parser = argparse.ArgumentParser(description="Card Payment (Authorize + Capture)")
    parser.add_argument(
        "--connector",
        required=True,
        help="Connector name (stripe, adyen, checkout, etc.)"
    )
    parser.add_argument(
        "--credentials",
        help="JSON file with connector credentials"
    )
    
    args = parser.parse_args()
    
    # Load credentials
    if args.credentials:
        with open(args.credentials) as f:
            credentials = json.load(f)
    else:
        credentials = {"api_key": "sk_test_dummy"}
        print("⚠️  Using dummy credentials. Set --credentials for real API calls.")
    
    # Run the flow
    result = asyncio.run(process_checkout_card(args.connector, credentials))
    
    print(f"\n{'='*60}")
    print(f"Result: {result}")
    print(f"{'='*60}")
    
    return 0 if result in ["success", "pending"] else 1


if __name__ == "__main__":
    exit(main())
# [END main]

File: examples/scenarios/checkout_card.rs
// Card Payment (Authorize + Capture) - Universal Example
//
// Works with any connector that supports card payments.
// Usage: cargo run --example checkout_card -- --connector stripe

use clap::Parser;
use grpc_api_types::payments::*;
use hyperswitch_payments_client::ConnectorClient;
use serde_json::json;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

// [START imports]
use std::error::Error;
// [END imports]

#[derive(Parser, Debug)]
#[command(name = "checkout_card")]
#[command(about = "Card Payment (Authorize + Capture) - Universal Example")]
struct Args {
    /// Connector name (stripe, adyen, checkout, etc.)
    #[arg(long)]
    connector: String,
    
    /// JSON file with connector credentials
    #[arg(long)]
    credentials: Option<String>,
}

// [START load_probe_data]
fn load_probe_data(connector_name: &str) -> Result<serde_json::Value, Box<dyn Error>> {
    let probe_path = format!("data/field_probe/{}.json", connector_name);
    let content = fs::read_to_string(&probe_path)?;
    let data: serde_json::Value = serde::from_str(&content)?;
    Ok(data)
}
// [END load_probe_data]

// [START stripe_config]
fn get_stripe_config(api_key: &str) -> ConnectorConfig {
    ConnectorConfig {
        connector: "stripe".to_string(),
        environment: Environment::Sandbox.into(),
        auth: ConnectorAuth::HeaderKey { 
            api_key: api_key.to_string().into() 
        },
        ..Default::default()
    }
}
// [END stripe_config]

// [START adyen_config]
fn get_adyen_config(api_key: &str, merchant_account: &str) -> ConnectorConfig {
    ConnectorConfig {
        connector: "adyen".to_string(),
        environment: Environment::Sandbox.into(),
        auth: ConnectorAuth::BodyKey { 
            api_key: api_key.to_string().into(),
            key1: merchant_account.to_string().into(),
        },
        ..Default::default()
    }
}
// [END adyen_config]

// [START checkout_config]
fn get_checkout_config(api_key: &str) -> ConnectorConfig {
    ConnectorConfig {
        connector: "checkout".to_string(),
        environment: Environment::Sandbox.into(),
        auth: ConnectorAuth::HeaderKey { 
            api_key: api_key.to_string().into() 
        },
        ..Default::default()
    }
}
// [END checkout_config]

// [START get_connector_config]
fn get_connector_config(connector_name: &str, credentials: &HashMap<String, String>) -> Result<ConnectorConfig, Box<dyn Error>> {
    match connector_name {
        "stripe" => Ok(get_stripe_config(
            credentials.get("api_key").ok_or("api_key required")?
        )),
        "adyen" => Ok(get_adyen_config(
            credentials.get("api_key").ok_or("api_key required")?,
            credentials.get("merchant_account").ok_or("merchant_account required")?
        )),
        "checkout" => Ok(get_checkout_config(
            credentials.get("api_key").ok_or("api_key required")?
        )),
        _ => Err(format!("Unknown connector: {}", connector_name).into()),
    }
}
// [END get_connector_config]

// [START build_authorize_request]
fn build_authorize_request(probe_data: &serde_json::Value, capture_method: &str) -> Result<PaymentServiceAuthorizeRequest, Box<dyn Error>> {
    let flows = probe_data.get("flows")
        .and_then(|f| f.get("authorize"))
        .ok_or("Connector doesn't support authorize flow")?;
    
    // Find Card or first supported payment method
    let mut card_data: Option<&serde_json::Value> = None;
    
    if let Some(obj) = flows.as_object() {
        for (pm_key, pm_data) in obj {
            if pm_data.get("status").and_then(|s| s.as_str()) == Some("supported") {
                if pm_key == "Card" {
                    card_data = Some(pm_data);
                    break;
                } else if card_data.is_none() {
                    card_data = Some(pm_data);
                }
            }
        }
    }
    
    let card_data = card_data.ok_or("No supported payment method found")?;
    let proto_request = card_data.get("proto_request")
        .ok_or("No proto_request in probe data")?;
    
    let mut request: PaymentServiceAuthorizeRequest = serde_json::from_value(proto_request.clone())?;
    request.capture_method = capture_method.to_string();
    
    Ok(request)
}
// [END build_authorize_request]

// [START build_capture_request]
fn build_capture_request(connector_transaction_id: &str, amount: &Amount) -> PaymentServiceCaptureRequest {
    PaymentServiceCaptureRequest {
        merchant_capture_id: "capture_001".to_string(),
        connector_transaction_id: connector_transaction_id.to_string(),
        amount_to_capture: Some(amount.clone()),
        ..Default::default()
    }
}
// [END build_capture_request]

// [START process_checkout_card]
pub async fn process_checkout_card(
    connector_name: &str,
    credentials: &HashMap<String, String>,
) -> Result<String, Box<dyn Error>> {
    println!("\n{}", "=".repeat(60));
    println!("Processing Card Payment via {}", connector_name);
    println!("{}", "=".repeat(60));
    
    // Load connector probe data
    let probe_data = load_probe_data(connector_name)?;
    
    // Build config
    let config = get_connector_config(connector_name, credentials)?;
    let client = ConnectorClient::new(config, None)?;
    
    // Build request from probe data
    let auth_request = build_authorize_request(&probe_data, "MANUAL")?;
    
    // Step 1: Authorize
    println!("\n[1/2] Authorizing...");
    let auth_response = client.authorize(auth_request, &HashMap::new(), None).await?;
    
    let status = auth_response.status();
    println!("Status: {:?}", status);
    println!("Connector Transaction ID: {:?}", auth_response.connector_transaction_id);
    
    // Handle status
    match status {
        PaymentStatus::Failure | PaymentStatus::AuthorizationFailed => {
            println!("❌ Payment declined");
            return Ok("failed".to_string());
        }
        PaymentStatus::Pending => {
            println!("⏳ Payment pending - awaiting async confirmation");
            println!("   (In production: wait for webhook, then poll get())");
            return Ok("pending".to_string());
        }
        PaymentStatus::Authorized => {
            println!("✅ Funds reserved successfully");
        }
        _ => {
            println!("⚠️  Unexpected status: {:?}", status);
            return Ok("error".to_string());
        }
    }
    
    // Step 2: Capture
    println!("\n[2/2] Capturing...");
    let capture_request = build_capture_request(
        auth_response.connector_transaction_id.as_deref().unwrap_or(""),
        auth_response.amount.as_ref().unwrap()
    );
    
    let capture_response = client.capture(capture_request, &HashMap::new(), None).await?;
    let capture_status = capture_response.status();
    println!("Capture Status: {:?}", capture_status);
    
    match capture_status {
        PaymentStatus::Captured | PaymentStatus::Authorized => {
            println!("✅ Payment captured successfully");
            Ok("success".to_string())
        }
        _ => {
            println!("⚠️  Capture returned: {:?}", capture_status);
            Ok("partial".to_string())
        }
    }
}
// [END process_checkout_card]

// [START main]
#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let args = Args::parse();
    
    // Load credentials
    let credentials: HashMap<String, String> = if let Some(creds_path) = args.credentials {
        let content = fs::read_to_string(creds_path)?;
        serde_json::from_str(&content)?
    } else {
        println!("⚠️  Using dummy credentials. Set --credentials for real API calls.");
        let mut map = HashMap::new();
        map.insert("api_key".to_string(), "sk_test_dummy".to_string());
        map
    };
    
    // Run the flow
    let result = process_checkout_card(&args.connector, &credentials).await?;
    
    println!("\n{}", "=".repeat(60));
    println!("Result: {}", result);
    println!("{}", "=".repeat(60));
    
    if result == "success" || result == "pending" {
        Ok(())
    } else {
        std::process::exit(1);
    }
}
// [END main]
